pkg/sfu/streamtracker/streamtracker_frame.go

```diff
@@ -0,0 +1,177 @@
+package streamtracker
+
+import (
+	"math"
+	"time"
+
+	"github.com/livekit/livekit-server/pkg/config"
+	"github.com/livekit/protocol/logger"
+)
+
+const (
+	checkInterval       = 500 * time.Millisecond
+	staleWindowFactor   = 5
+	frameRateResolution = float64(0.01) // 1 frame every 100 seconds
+)
+
+type StreamTrackerFrameParams struct {
+	Config    config.StreamTrackerFrameConfig
+	ClockRate uint32
+	Logger    logger.Logger
+}
+
+type StreamTrackerFrame struct {
+	params StreamTrackerFrameParams
+
+	initialized bool
+
+	tsInitialized bool
+	oldestTS      uint32
+	newestTS      uint32
+	numFrames     int
+
+	lowestFrameRate   float64
+	evalInterval      time.Duration
+	lastStatusCheckAt time.Time
+}
+
+func NewStreamTrackerFrame(params StreamTrackerFrameParams) StreamTrackerImpl {
+	s := &StreamTrackerFrame{
+		params: params,
+	}
+	s.Reset()
+	return s
+}
+
+func (s *StreamTrackerFrame) Start() {
+}
+
+func (s *StreamTrackerFrame) Stop() {
+}
+
+func (s *StreamTrackerFrame) Reset() {
+	s.initialized = false
+
+	s.tsInitialized = false
+	s.oldestTS = 0
+	s.newestTS = 0
+	s.numFrames = 0
+
+	s.lowestFrameRate = 0.0
+	s.updateEvalInterval()
+	s.lastStatusCheckAt = time.Time{}
+}
+
+func (s *StreamTrackerFrame) GetCheckInterval() time.Duration {
+	return checkInterval
+}
+
+func (s *StreamTrackerFrame) Observe(hasMarker bool, ts uint32) StreamStatusChange {
+	if !s.initialized {
+		s.initialized = true
+		if hasMarker {
```

Comment:
I had this like
```
        // Start only on a key frame. This condition should hold true as this path is exercised under following conditions                                                 
        //   1. When track starts                                                                                                                                          
        //   2. When track resumes after an unmute.                                                                                                                        
        //   3. When a layer starts after dynacast pause.                                                                                                                  
        if isKeyFrame {                                                                                                                                                    
            s.initialized = true                                                                                                                                           
            if hasMarker {                                                                                                                                                 
                s.tsInitialized = true                                                                                                                                     
                s.oldestTS = ts                                                                                                                                            
                s.newestTS = ts                                                                                                                                            
                s.numFrames = 1                                                                                                                                            
            }                                                                                                                                                              
            return StreamStatusChangeActive                                                                                                                                
        }
```
But, it turned out to be not true. I thought browsers will use a key frame if the previous frame is too far back. But, they don't. If I disable camera, wait for 15 seconds and then re-start the camera, it sends a P frame with the old frame as reference. Makes sense to make a P frame if the old frame is in the encoder memory. So, removed that and just using re-start of traffic. This is effectively packet based, i. e. declaring active on first packet just like the packet based stream tracker works.